File: maxon-zig/src/compiler/ast.zig
const std = @import("std");

pub const Program = struct {
    types: []TypeDecl,
    enums: []EnumDecl,
    interfaces: []InterfaceDecl,
    functions: []FunctionDecl,
};

pub const InterfaceMethod = struct {
    name: []const u8,
    is_static: bool,
    params: []const ParamDecl,
    return_type: ?TypeExpr, // null for void
    has_default_impl: bool,
    default_body: ?[]Statement,
};

pub const InterfaceDecl = struct {
    name: []const u8,
    is_export: bool,
    generic_params: []const []const u8, // ["Element"] for `uses Element`
    extends: []const []const u8, // ["Collection", "Iterable"] for `extends Collection, Iterable`
    methods: []InterfaceMethod,
};

pub const EnumDecl = struct {
    name: []const u8,
    members: []const []const u8,
};

pub const InterfaceConformance = struct {
    interface_name: []const u8,
    type_args: []const []const u8,
};

pub const MethodDecl = struct {
    name: []const u8,
    qualified_name: ?[]const u8, // "Collection.count" for interface methods
    is_static: bool,
    is_export: bool,
    params: []const ParamDecl,
    return_type: ?TypeExpr, // null for void
    body: []Statement,
};

pub const TypeDecl = struct {
    name: []const u8,
    is_export: bool,
    generic_params: []const []const u8, // ["Element"] for `uses Element`
    conformances: []const InterfaceConformance,
    fields: []FieldDecl,
    methods: []MethodDecl,
};

pub const FieldDecl = struct {
    name: []const u8,
    type_expr: TypeExpr,
    is_mutable: bool,
};

pub const ArrayTypeExpr = struct {
    size: ?i64, // null for unsized (array of int), value for sized (array of 3 int)
    element_type: []const u8,
};

pub const GenericTypeExpr = struct {
    base_type: []const u8, // Array, Map, etc.
    type_args: []const []const u8, // [int], [string, int], etc.
};

pub const TypeExpr = union(enum) {
    simple: []const u8, // int, float, MyStruct
    array: ArrayTypeExpr, // array of int, array of 3 int
    generic: GenericTypeExpr, // Array of int, Map of string int
    optional: *const TypeExpr, // T or nil
};

pub const ParamDecl = struct {
    name: []const u8,
    type_expr: TypeExpr,
};

pub const FunctionDecl = struct {
    name: []const u8,
    is_export: bool,
    params: []const ParamDecl,
    return_type: ?TypeExpr, // null for void
    body: []Statement,
};

pub const IndexAssign = struct {
    base: *const Expression,
    index: *const Expression,
    value: Expression,
};

pub const AssignStmt = struct {
    target: []const u8,
    value: Expression,
};

pub const FieldAssign = struct {
    base: *const Expression,
    field_name: []const u8,
    value: Expression,
};

pub const IfStmt = struct {
    condition: Expression,
    body: []Statement,
    label: []const u8,
    else_body: ?[]Statement, // For else clause
    else_label: ?[]const u8, // Label for else block
    else_if: ?*const IfStmt, // For else-if chain (recursive)

    // If-let binding (optional): if let name = expr 'label'
    binding_name: ?[]const u8 = null,
};

pub const WhileStmt = struct {
    condition: Expression,
    body: []Statement,
    label: []const u8,
};

pub const ForStmt = struct {
    var_name: []const u8, // loop variable name
    iterable: Expression, // expression to iterate over
    body: []Statement,
    label: []const u8,
};

pub const BreakStmt = struct {};
pub const ContinueStmt = struct {};

pub const ElseUnwrapDecl = struct {
    var_name: []const u8,
    optional_expr: *const Expression,
    default_body: []Statement,
    label: []const u8,
};

pub const Statement = union(enum) {
    @"return": ReturnStmt,
    let_decl: VarDecl,
    var_decl: VarDecl,
    index_assign: IndexAssign,
    assign: AssignStmt,
    field_assign: FieldAssign,
    call: CallExpr,
    method_call: MethodCallExpr,
    if_stmt: IfStmt,
    while_stmt: WhileStmt,
    for_stmt: ForStmt,
    break_stmt: BreakStmt,
    continue_stmt: ContinueStmt,
    else_unwrap_decl: ElseUnwrapDecl,
};

pub const VarDecl = struct {
    name: []const u8,
    type_annotation: ?TypeExpr, // Optional explicit type
    value: Expression,
};

pub const ReturnStmt = struct {
    value: ?Expression,
};

pub const BinaryOp = enum {
    add,
    sub,
    mul,
    div,
    mod,
};

pub const CompareOp = enum {
    eq,
    ne,
    lt,
    le,
    gt,
    ge,
};

pub const LogicalOp = enum {
    @"and",
};

pub const LogicalExpr = struct {
    left: *const Expression,
    op: LogicalOp,
    right: *const Expression,
};

pub const UnaryOp = enum {
    negate, // -x
    not, // not x (for future use)
};

pub const UnaryExpr = struct {
    op: UnaryOp,
    operand: *const Expression,
};

pub const BinaryExpr = struct {
    left: *const Expression,
    op: BinaryOp,
    right: *const Expression,
};

pub const CompareExpr = struct {
    left: *const Expression,
    op: CompareOp,
    right: *const Expression,
};

pub const CallExpr = struct {
    func_name: []const u8,
    args: []const Expression,
};

pub const FieldInit = struct {
    name: []const u8,
    value: *const Expression,
};

pub const StructInitExpr = struct {
    type_name: []const u8,
    type_args: []const []const u8, // ["int"] for `Container of int{...}`
    fields: []const FieldInit,
};

pub const FieldAccessExpr = struct {
    base: *const Expression,
    field_name: []const u8,
};

pub const ArrayLiteralExpr = struct {
    elements: []const Expression,
};

pub const IndexExpr = struct {
    base: *const Expression,
    index: *const Expression,
};

pub const SizedArrayExpr = struct {
    size: *const Expression,
    element_type: []const u8,
};

pub const MethodCallExpr = struct {
    base: *const Expression,
    method_name: []const u8,
    args: []const Expression,
};

pub const NilCoalesceExpr = struct {
    optional: *const Expression,
    default: *const Expression,
};

pub const Expression = union(enum) {
    integer: i64,
    float_lit: f64,
    bool_lit: bool,
    nil_lit,
    self_expr, // reference to current instance in methods
    identifier: []const u8,
    unary: UnaryExpr,
    binary: BinaryExpr,
    compare: CompareExpr,
    logical: LogicalExpr,
    call: CallExpr,
    struct_init: StructInitExpr,
    field_access: FieldAccessExpr,
    array_literal: ArrayLiteralExpr,
    index: IndexExpr,
    sized_array: SizedArrayExpr,
    method_call: MethodCallExpr,
    nil_coalesce: NilCoalesceExpr,
};

File: maxon-sharp/Compiler/AST.cs
namespace MaxonSharp.Compiler;

// ============================================================================
// Source Span
// ============================================================================

public record SourceLocation(int FileIndex, int StartLine, int StartColumn, int EndLine, int EndColumn, string? Identifier = null) {
	public int Line => StartLine;
	public int Column => StartColumn;
	public static SourceLocation Point(int fileIndex, int line, int column) =>
		new(fileIndex, line, column, line, column, null);

	public static SourceLocation Block(int fileIndex, int startLine, int startColumn, int endLine, string? identifier) =>
		new(fileIndex, startLine, startColumn, endLine, 0, identifier);
}

// ============================================================================
// Type References
// ============================================================================

public abstract record TypeRef {
	public required SourceLocation Location { get; init; }
}

public record SimpleTypeRef(string Name) : TypeRef;
public record GenericTypeRef(string BaseName, List<string> TypeArgs) : TypeRef;
public record FunctionTypeRef(List<TypeRef> ParamTypes, TypeRef? ReturnType) : TypeRef;
public record ErrorUnionTypeRef(TypeRef SuccessType, string ErrorType) : TypeRef;

// ============================================================================
// Expressions
// ============================================================================

public abstract record Expr {
	public required SourceLocation Location { get; init; }
}

// Literals
public record IntLiteralExpr(long Value) : Expr;
public record FloatLiteralExpr(double Value) : Expr;
public record BoolLiteralExpr(bool Value) : Expr;
public record StringLiteralExpr(string Value) : Expr;
public record CharLiteralExpr(string Value) : Expr;
public record IdentifierExpr(string Name) : Expr;
public record SelfExpr : Expr;

// Interpolated string: "Hello {name}!"
public record InterpolatedPart(bool IsExpression, string? LiteralValue, Expr? Expression, string? FormatSpec);
public record InterpolatedStringExpr(List<InterpolatedPart> Parts) : Expr;

// Operators
public enum BinaryOp { Add, Sub, Mul, Div, Mod, Band, Bor, Bxor, Shl, Shr }
public enum CompareOp { Eq, Ne, Lt, Le, Gt, Ge }
public enum LogicalOp { And, Or }
public enum UnaryOp { Negate, Not }

public record BinaryExpr(Expr Left, BinaryOp Op, Expr Right) : Expr;
public record CompareExpr(Expr Left, CompareOp Op, Expr Right) : Expr;
public record LogicalExpr(Expr Left, LogicalOp Op, Expr Right) : Expr;
public record UnaryExpr(UnaryOp Op, Expr Operand) : Expr;

// Calls and member access
public record NamedArg(string Name, Expr Value);

public enum BuiltinOp {
	None = 0,
	Trunc,
	Sqrt,
	Floor,
	Ceil,
	Round,
	Abs,
	Min,
	Max
}

public record CallExpr(string FuncName, List<Expr> Args, List<NamedArg> NamedArgs, BuiltinOp Builtin = BuiltinOp.None) : Expr;
public record MethodCallExpr(Expr Base, string MethodName, List<Expr> Args, List<NamedArg> NamedArgs) : Expr;
public record FieldAccessExpr(Expr Base, string FieldName) : Expr;
public record IndexExpr(Expr Base, Expr Index) : Expr;

// Struct and array literals
public record FieldInit(string Name, Expr Value);
public record StructInitExpr(string? TypeName, List<string> TypeArgs, List<FieldInit> Fields) : Expr;
public record ArrayLiteralExpr(List<Expr> Elements) : Expr;

// Map literal
public record MapEntry(Expr Key, Expr Value);
public record MapLiteralExpr(List<MapEntry> Entries) : Expr;

// Static method call: Type.method(args) - resolved to enum case or static method in HIR
public record StaticCallExpr(string TypeName, string MemberName, List<Expr> Args, List<NamedArg> NamedArgs) : Expr;

// Enum case construction: Result.success(42) - generated from StaticCallExpr in HIR
public record EnumCaseExpr(string EnumName, string CaseName, List<Expr> Args) : Expr;

// Cast expression: value as int
public record CastExpr(Expr Expression, string TargetType) : Expr;

// Error handling: try expr otherwise default
public enum OtherwiseMode { DefaultExpr, Ignore, Block, BlockWithErr }
public record OtherwiseClause(OtherwiseMode Mode, Expr? DefaultExpr, string? ErrorBinding, List<Stmt>? Body);
public record TryExpr(Expr Expression, OtherwiseClause? Otherwise) : Expr;

// Match expression
public record PatternBinding(string CaseName, List<string> Bindings);
public record MatchExprCase(List<Expr> Patterns, List<PatternBinding?> PatternBindings, Expr Result);
public record MatchExpr(Expr Scrutinee, List<MatchExprCase> Cases, Expr? DefaultExpr, string Label) : Expr;

// Range expression for iteration: start..end, start..<end, start..=end
public record RangeExpr(Expr Start, Expr End, bool Inclusive) : Expr;

// Range pattern for match: 1..=5, 1..<5, 1.., ..=5
public record RangePatternExpr(Expr? Lower, Expr? Upper, bool Inclusive) : Expr;

// Closure: (x int) gives x * 2
public record ClosureParam(string Name, string? TypeName);
public record ClosureExpr(List<ClosureParam> Params, Expr Body) : Expr;

// InitFromArray: TypeName from [1, 2, 3]
public record InitFromArrayExpr(string TypeName, List<string> TypeArgs, Expr Elements) : Expr;

// ============================================================================
// Statements
// ============================================================================

public abstract record Stmt {
	public required SourceLocation Location { get; init; }
}

public record ReturnStmt(Expr? Value) : Stmt;
public record LetDeclStmt(string Name, TypeRef? TypeAnnotation, Expr Value) : Stmt;
public record VarDeclStmt(string Name, TypeRef? TypeAnnotation, Expr Value) : Stmt;
public record AssignStmt(string Target, Expr Value) : Stmt;
public record FieldAssignStmt(Expr Base, string FieldName, Expr Value) : Stmt;
public record IndexAssignStmt(Expr Base, Expr Index, Expr Value) : Stmt;
public record ExprStmt(Expr Expression) : Stmt;  // For call/method_call as statement
public record ThrowStmt(Expr ErrorExpr) : Stmt;
public record BreakStmt(string? Label) : Stmt;
public record ContinueStmt(string? Label) : Stmt;

// Control flow with child blocks

public record IfStmt(Expr Condition, List<Stmt> ThenBody, SourceLocation ThenBlock, List<Stmt>? ElseBody, SourceLocation? ElseBlock) : Stmt;
public record WhileStmt(Expr Condition, List<Stmt> Body, SourceLocation Block) : Stmt;
public record ForStmt(string VarName, Expr Iterable, List<Stmt> Body, SourceLocation Block) : Stmt;

// Match statement
public record MatchCase(List<Expr> Patterns, List<PatternBinding?> PatternBindings, List<Stmt> Body, bool HasFallthrough);
public record MatchStmt(Expr Scrutinee, List<MatchCase> Cases, List<Stmt>? DefaultBody, SourceLocation Block) : Stmt;

// ============================================================================
// Declarations
// ============================================================================

public record ParamDecl(string Name, TypeRef Type, Expr? DefaultValue = null) {
	public required SourceLocation Location { get; init; }
}

public record FieldDecl(string Name, TypeRef? Type, bool IsMutable, bool IsExport = false, bool IsStatic = false, Expr? DefaultValue = null) {
	public required SourceLocation Location { get; init; }
}

public record MethodDecl(
	string Name,
	bool IsStatic,
	bool IsExport,
	List<ParamDecl> Params,
	TypeRef? ReturnType,
	string? ThrowsType,
	List<Stmt> Body,
	SourceLocation Block,
	string? DocComment = null
);

public record FunctionDecl(
	string Name,
	bool IsExport,
	List<ParamDecl> Params,
	TypeRef? ReturnType,
	string? ThrowsType,
	List<Stmt> Body,
	SourceLocation Block,
	string? DocComment = null
);

public record TypeDecl(
	string Name,
	bool IsExport,
	List<string> GenericParams,
	List<InterfaceConformance> Conformances,
	List<TypeAliasDecl> AssociatedTypes,
	List<FieldDecl> Fields,
	List<MethodDecl> Methods,
	SourceLocation Block
) {
	public required SourceLocation Location { get; init; }
}

public record EnumMember(string Name, Expr? Value, List<ParamDecl> AssociatedValues) {
	public required SourceLocation Location { get; init; }
}

public record EnumDecl(
	string Name,
	bool IsExport,
	List<InterfaceConformance> Conformances,
	List<EnumMember> Members,
	List<MethodDecl> Methods,
	SourceLocation Block
);

public record InterfaceConformance(string InterfaceName, List<string> TypeArgs);

public record InterfaceMethod(string Name, bool IsStatic, List<ParamDecl> Params, TypeRef? ReturnType, string? ThrowsType);

public record InterfaceDecl(
	string Name,
	bool IsExport,
	List<string> GenericParams,
	List<string> Extends,
	List<TypeAliasDecl> AssociatedTypes,
	List<InterfaceMethod> Methods,
	SourceLocation Block
);

public record ExtensionMethod(
	string Name,
	List<ParamDecl> Params,
	TypeRef? ReturnType,
	string? ThrowsType,
	List<Stmt> Body,
	SourceLocation Block
);

public record ExtensionDecl(
	string InterfaceName,
	bool IsExport,
	List<TypeAliasDecl> AssociatedTypes,
	List<ExtensionMethod> Methods,
	SourceLocation Block
);

public record TypeAliasDecl(string Name, string BaseType, List<string> TypeArgs, bool IsExport = false) {
	public required SourceLocation Location { get; init; }
}

public record GlobalConstant(string Name, bool IsExport, Expr Value) {
	public required SourceLocation Location { get; init; }
}

public record GlobalVariable(string Name, bool IsExport, Expr Value) {
	public required SourceLocation Location { get; init; }
}

// ============================================================================
// Program (root)
// ============================================================================

public record ProgramAst(
	List<TypeDecl> Types,
	List<EnumDecl> Enums,
	List<InterfaceDecl> Interfaces,
	List<ExtensionDecl> Extensions,
	List<FunctionDecl> Functions,
	List<GlobalConstant> GlobalConstants,
	List<GlobalVariable> GlobalVariables,
	List<TypeAliasDecl> TypeAliases
) {
	/// <summary>
	/// Merges multiple ProgramAst objects into a single unified program.
	/// File scoping is enforced during semantic analysis.
	/// </summary>
	/// <param name="programs">List of parsed programs to merge</param>
	/// <param name="mainFileIndex">Index of the main file (reserved for future use)</param>
	public static ProgramAst Merge(List<ProgramAst> programs) {
		var types = new List<TypeDecl>();
		var enums = new List<EnumDecl>();
		var interfaces = new List<InterfaceDecl>();
		var extensions = new List<ExtensionDecl>();
		var functions = new List<FunctionDecl>();
		var globalConstants = new List<GlobalConstant>();
		var globalVariables = new List<GlobalVariable>();
		var typeAliases = new List<TypeAliasDecl>();

		foreach (var p in programs) {
			types.AddRange(p.Types);
			enums.AddRange(p.Enums);
			interfaces.AddRange(p.Interfaces);
			extensions.AddRange(p.Extensions);
			functions.AddRange(p.Functions);
			globalConstants.AddRange(p.GlobalConstants);
			globalVariables.AddRange(p.GlobalVariables);
			typeAliases.AddRange(p.TypeAliases);
		}

		return new ProgramAst(types, enums, interfaces, extensions, functions, globalConstants, globalVariables, typeAliases);
	}
}
